packages/sdk-client/test/client.test/sdk-client.test.ts

```diff
@@ -523,3 +525,372 @@ describe('process', () => {
       })
   })
 })
+
+// exposed process function
+describe('process - exposed', () => {
+  const request = {
+    uri: '/test/products',
+    method: 'GET',
+    body: null,
+    headers: {},
+  } as any
+
+  describe('validate arguments', () => {
+    const middlewares = [
+      (next) =>
+        (...args) =>
+          next(...args),
+    ]
+
+    createClient({ middlewares }) as any
+    test('should throw if second argument missing', () => {
+      expect(() => Process(request, null, {})).toThrow(
```

Comment:
Why `Process` is a separate function and not like this?
```
await apiRoot
  .withProjectKey({ projectKey: '{projectKey}' })
  .customers()
  .get()
  .process((response) => doSth(response))
```
I saw Jen's comment, but I do not understand what is the problem if it is like this? IMO it's more intuitive than to have a special function `Process`. Thanks.

Reply:
Cause process function embedded can be misleading. Not every request may be suitable to be processed as they can be inefficient.

Reply:
And process only makes sense for queries and not like every method at every endpoint across all APIs

Reply:
Thanks, it makes sense, but how a static `Process` method will help in this case? Do you plan it to be like a 'hidden' feature that only if a user knows about it, they will use it?

> And process only makes sense for queries and not like every method at every endpoint across all APIs
Could TS help in this case so that `process` would be available only for queries?

Reply:
I think we will have to add some kind of interface in the future in order to limit it correctly. As like as we did for the java SDK.

It should not be a hidden feature, but it must be documented and promoted. The exposed process function should make it easier to use it without the need to have a client instance explicitly.

Reply:
From a user's perspective it would be much easier to have an embedded function. We had it in the [old SDK](https://github.com/commercetools/sphere-node-sdk/blob/096ae225bf501103d679cda034b20d92ab31036f/src/coffee/services/base.coffee#L366) and it was working fine. 
But if you think there is a problem with the approach, what you have now is also ok for me.

Reply:
The only way to do it like [this](https://github.com/commercetools/sphere-node-sdk/blob/096ae225bf501103d679cda034b20d92ab31036f/src/coffee/services/base.coffee#L366) with our current setup is to add `a process` method in the `ApiRoot` class, which as Jens said is not the best thing to do.

For now, there are just two options that I can think of:

- Create a `client` and call the `process` method on the client just as we mentioned before
- Use this embedded `Process` function.

One other option would be to dynamically add the process method to the ApiRoot at runtime, but again we running from polluting the class and also this might introduce some unexpected behaviour.

Reply:
The linked old SDK just covered the platform API and introduced a Service concept for each domain. The newer SDKs are separating the client and the requests. This gives the opportunity to reuse the client configuration, as well as the requests itself and add middlewares much simpler. Also it's nearer to the nature of an HTTP api.

The request builders (api roots) are more of an helper to avoid memorizing all endpoints, parameters etc and to code exploratively. And it avoids breaking changes if new parameters are getting introduces or some names may change as the structure is not changing